vcd/resource_vcd_cse_kubernetes_cluster.go

```diff
@@ -0,0 +1,796 @@
+package vcd
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"github.com/hashicorp/go-cty/cty"
+	semver "github.com/hashicorp/go-version"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+	"sort"
+	"time"
+)
+
+func resourceVcdCseKubernetesCluster() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceVcdCseKubernetesClusterCreate,
+		ReadContext:   resourceVcdCseKubernetesRead,
+		UpdateContext: resourceVcdCseKubernetesUpdate,
+		DeleteContext: resourceVcdCseKubernetesDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceVcdCseKubernetesImport,
+		},
+		Schema: map[string]*schema.Schema{
+			"cse_version": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"4.1.0", "4.1.1", "4.2.0"}, false),
+				Description:  "The CSE version to use",
+				DiffSuppressFunc: func(k, oldValue, newValue string, d *schema.ResourceData) bool {
+					// This custom diff function allows to correctly compare versions.
+					oldVersion, err := semver.NewVersion(oldValue)
+					if err != nil {
+						return false
+					}
+					newVersion, err := semver.NewVersion(newValue)
+					if err != nil {
+						return false
+					}
+					return oldVersion.Equal(newVersion)
+				},
+				DiffSuppressOnRefresh: true,
+			},
+			"runtime": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "tkg",
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"tkg"}, false), // May add others in future releases of CSE
+				Description:  "The Kubernetes runtime for the cluster. Only 'tkg' (Tanzu Kubernetes Grid) is supported",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The name of the Kubernetes cluster",
+				ValidateDiagFunc: matchRegex(`^[a-z](?:[a-z0-9-]{0,29}[a-z0-9])?$`, "name must contain only lowercase alphanumeric characters or '-',"+
+					"start with an alphabetic character, end with an alphanumeric, and contain at most 31 characters"),
```

Comment:
A query of caution - any risk of overvalidating it (e.g. someone has a cluster that is named different and cannot import it because name doesn't match constraint)? Would it be rough if we left that validation for VCD API part?

Reply:
This is the same validation for UI, so if that happens then we have a cheater 😈 

More seriously, I don't think we should leave it to VCD to validate, because it won't validate. I mean, you send the wrong name in the RDE JSON, the RDE gets created (because the endpoint doesn't care about the payload), then wait X minutes for the CSE Server to realize that the cluster is invalid and then finish the creation process as `error` (that's if `auto_repair_on_errors=false`, otherwise it will try again and again).

So, I see your point. I'd agree if this was a more straight forward API endpoint, that immediately returns feedback, but in this case it's more of a "black hole" that swallows your payload and then you wait and wait for something to happen. I think giving an early feedback is more convenient (this validation actually saved a few manual tests from failing because I'm used to `camelCase` 😄 )

Reply:
> More seriously, I don't think we should leave it to VCD to validate, because it won't validate.

This answers my query :)